Transient navigators

aengelberg opened the pull request:
A suite of navigators that work on transient collections is available in the `com.rpl.specter.transient` namespace.

For transient vectors: `END!`, `FIRST!`, `LAST!`, `keypath!`
For transient maps: `keypath!`, `submap!`
For transient sets: none supported

aengelberg commented:
I didn't include `subset!` because the operation `(transform [(subset! #{large set})] (transient #{small set}))` would have to do a linear pass all the way through that large set unlike the persistent version which uses `set/intersection`. If there is some feature I'm not seeing in the transient API, or that performance issue isn't a dealbreaker, I'm open to adding it.

aengelberg commented:
Currently `END!` actually does work on transient maps and sets (it passes on `[]` then dutifully dumps the result into the respective collection). My current thinking is to not declare (via docstrings, etc) that those collections are supported, because 

1. the immutable `END` does not support maps or sets
2. the name "end" would be confusing if it supported non-sequential structures

The support for map is less of a concern because `(submap! [])` is an option. I'm more concerned about sets because no other transient operations are available for them. Thoughts?

Ready for this pull request to be reviewed.

nathanmarz commented:
The code looks good. Only one thing – I've gotten much more rigorous now about tracking and verifying Specter's performance, so I'd like to see benchmarks added to benchmarks.clj for these new features. This will prevent performance regressions and give us targets for further optimization.

nathanmarz commented on src/clj/com/rpl/specter/transient.cljx:
```diff
@@ -0,0 +1,91 @@
+        (reduce (fn [m k]
+                  (dissoc! m k))
+                % m-keys)
+        (reduce (fn [m [k v]]
```
`reduce-kv` will be faster here

aengelberg commented:
Sounds good re: benchmarks. What would you primarily like to compare; the difference of transient navigators to persistent navigators, or transient navigators to specter-less transient operations? I'm less sure how to do the latter, because built in transient functions of any reasonable complexity simply don't exist (no "assoc!-in", etc)

nathanmarz commented:
I'd say to compare both. The latter will give us an upper-bound on how fast these navigators could possibly be. To do those, just compare the transient operation against the Specter equivalent, e.g.:

```clojure
(setval END! [1] atransient)
(conj! atransient2 1)
```

aengelberg commented:
Addressed your comments.

FYI, after adding the benchmarks I noticed that `submap` was only just a smidge slower than `submap!` (x1.05). After I changed the instance of `reduce` to `reduce-kv` as you suggested, that factor went up to about x1.20. Open to suggestions on how to improve that benchmark test (not sure what scenarios `submap` is usually used).

nathanmarz commented:
I merged too fast – after merging I saw some updates necessary to the benchmarks. It was pretty fast to do the updates so I just did it myself: https://github.com/nathanmarz/specter/commit/82321d73707fbbd0421de1bd27fe87d0a0a1d95b Let me know if anything there seems amiss.

I noticed from the benchmark that the `keypath` and `keypath!` navigators have lots of room for improved performance – #24 I believe would close most of that gap.

By the way, if a path only contains a single navigator you can omit the '[...]' around it.
